RecoCTPPS/TotemRPLocal/plugins/TotemRPLocalTrackFitter.cc

```diff
@@ -0,0 +1,185 @@
+
+    virtual ~TotemRPLocalTrackFitter() {}
+
+    virtual void produce(edm::Event& e, const edm::EventSetup& c);
+
+  private:
+    int verbosity_;
+
+    /// Selection of the pattern-recognition module.
+    edm::InputTag tagUVPattern;
+
+    edm::EDGetTokenT<edm::DetSetVector<TotemRPUVPattern>> patternCollectionToken;
+    
+    /// A watcher to detect geometry changes.
+    edm::ESWatcher<VeryForwardRealGeometryRecord> geometryWatcher;
+    
+    /// The instance of the fitter module
+    TotemRPLocalTrackFitterAlgorithm fitter_;
+};
+
+//----------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------
+
+using namespace std;
+using namespace edm;
+
+//----------------------------------------------------------------------------------------------------
+
+TotemRPLocalTrackFitter::TotemRPLocalTrackFitter(const edm::ParameterSet& conf)
+   : verbosity_(conf.getParameter<int>("verbosity")), fitter_(conf)
+{
+  tagUVPattern = conf.getParameter<edm::InputTag>("tagUVPattern");
+  patternCollectionToken = consumes<DetSetVector<TotemRPUVPattern>>(tagUVPattern);
+
+  produces<DetSetVector<TotemRPLocalTrack>>();
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void TotemRPLocalTrackFitter::produce(edm::Event& e, const edm::EventSetup& setup)
+{
+  if (verbosity_ > 5)
+    printf(">> TotemRPLocalTrackFitter::produce\n");
+
+  // get geometry
+  edm::ESHandle<TotemRPGeometry> geometry;
+  setup.get<VeryForwardRealGeometryRecord>().get(geometry);
+
+  if (geometryWatcher.check(setup))
+    fitter_.Reset();
+  
+  // get input
+  edm::Handle<DetSetVector<TotemRPUVPattern>> input;
+  e.getByToken(patternCollectionToken, input);
+
+  // run fit for each RP
+  DetSetVector<TotemRPLocalTrack> output;
+  
+  for (const auto &rpv : *input)
+  {
+    det_id_type rpId =  rpv.detId();
+
+    // is U-V association unique?
+    unsigned int n_U=0, n_V=0;
+    unsigned int idx_U=0, idx_V=0;
+    for (unsigned int pi = 0; pi < rpv.size(); pi++)
+    {
+      const TotemRPUVPattern &pattern = rpv[pi];
+
+      // here it would make sense to skip non-fittable patterns, but to keep the logic
+      // equivalent to version 7_0_4, nothing is skipped
+      /*
+      if (pattern.getFittable() == false)
+        continue;
+      */
+
+      switch (pattern.getProjection())
+      {
+        case TotemRPUVPattern::projU:
+          n_U++;
+          idx_U=pi;
+          break;
+     
+        case TotemRPUVPattern::projV:
+          n_V++;
+          idx_V=pi;
+          break;
+
+        default:
+          break;
+      }   
+    }
+
+    if (n_U != 1 || n_V != 1)
+    {
+      if (verbosity_)
+        printf(">> TotemRPLocalTrackFitter::produce > Impossible to combine U and V patterns in RP %u (n_U=%u, n_V=%u).\n",
```

Comment:
Using `printf` even for debugging levels isn't a good idea. If you run with more than one thread all your output will get garbled. The message logger is designed to handle multiple threads.